reactor-netty5-http/src/main/java/reactor/netty5/http/client/HttpClientFormEncoder.java

```diff
@@ -0,0 +1,356 @@
+			if (contentType == null) {
+				scontentType = DEFAULT_BINARY_CONTENT_TYPE;
+			}
+			MemoryFileUpload fileUpload = new MemoryFileUpload(name,
+					filename,
+					scontentType,
+					DEFAULT_TRANSFER_ENCODING,
+					newCharset,
+					-1);
+			fileUpload.setMaxSize(-1);
+			fileUpload.setContent(stream);
+			addBodyHttpData(fileUpload);
+		}
+		catch (ErrorDataEncoderException e) {
+			throw Exceptions.propagate(e);
+		}
+		catch (IOException e) {
+			throw Exceptions.propagate(new ErrorDataEncoderException(e));
+		}
+		return this;
+	}
+
+	@Override
+	public HttpClientForm files(String name,
+			File[] files,
+			String[] contentTypes) {
+		for (int i = 0; i < files.length; i++) {
+			file(name, files[i], contentTypes[i]);
+		}
+		return this;
+	}
+
+	@Override
+	public HttpClientForm files(String name,
+			File[] files,
+			String[] contentTypes,
+			boolean[] textFiles) {
+		try {
+			addBodyFileUploads(name, files, contentTypes, textFiles);
+		}
+		catch (ErrorDataEncoderException e) {
+			throw Exceptions.propagate(e);
+		}
+		return this;
+	}
+
+	@Override
+	public HttpClientForm encoding(EncoderMode mode) {
+		this.newMode = Objects.requireNonNull(mode, "mode");
+		this.needNewEncoder = true;
+		return this;
+	}
+
+	@Override
+	public HttpClientForm multipart(boolean isMultipart) {
+		this.needNewEncoder = isChunked() != isMultipart;
+		this.newMultipart = isMultipart;
+		return this;
+	}
+
+	@Override
+	public HttpClientForm textFile(String name, File file) {
+		textFile(name, file, null);
+		return this;
+	}
+
+	@Override
+	public HttpClientForm textFile(String name, File file, @Nullable String contentType) {
+		try {
+			addBodyFileUpload(name, file, contentType, true);
+		}
+		catch (ErrorDataEncoderException e) {
+			throw Exceptions.propagate(e);
+		}
+		return this;
+	}
+
+	@Override
+	public HttpClientForm textFile(String name, InputStream stream) {
+		textFile(name, stream, null);
+		return this;
+	}
+
+	@Override
+	public HttpClientForm textFile(String name,
+			InputStream stream,
+			@Nullable String contentType) {
+		Objects.requireNonNull(name, "name");
+		Objects.requireNonNull(stream, "stream");
+		try {
+			String scontentType = contentType;
+
+			if (contentType == null) {
+				scontentType = DEFAULT_TEXT_CONTENT_TYPE;
+			}
+
+			MemoryFileUpload fileUpload =
+					new MemoryFileUpload(name, "", scontentType, null, newCharset, -1);
+			fileUpload.setMaxSize(-1);
+			fileUpload.setContent(stream);
+			addBodyHttpData(fileUpload);
+		}
+		catch (ErrorDataEncoderException e) {
+			throw Exceptions.propagate(e);
+		}
+		catch (IOException e) {
+			throw Exceptions.propagate(new ErrorDataEncoderException(e));
+		}
+		return this;
+	}
+
+	@Override
+	public void run() {
+		cleanFiles();
+		// Clean list of all InterfaceHttpData from body part
+		for (InterfaceHttpData bodyAttribute : getBodyListAttributes()) {
+			if (bodyAttribute.isAccessible()) {
+				bodyAttribute.close();
+			}
+		}
```

Comment:
Please elaborate why this is needed.

Reply:
This avoids a memory leak which also happens in 1.0.x (if correct then this needs to be first committed in 1.0.x, then forward merged here ...).

So, when an http **file** upload is added using [HttpClientFormEncoder.file](https://github.com/reactor/reactor-netty/pull/2327/files#diff-e6fc47c2c0ba258df7d2d4bd4bd71cb91406d066a3130eadde1f70482cb53356R195
) method **line 195**, then the _MemoryFileUpload_ object is instantiated, but not created using the _newFactory_ factory field (probably for good reasons, I don't know). So the _MemoryFileUpload_ is not stored in the _DefaultHttpDataFactory_ data structure (in _DefaultHttpDataFactory.requestFileDeleteMap_), but is only stored in the _HttpPostRequestEncoder.bodyListDatas_ list.

This means that if we only call cleanFiles(), we will only clean the MixAttributes from DefaultHttpDataFactory.requestFileDeleteMap, but the _MemoryFileUpload_ from the _HttpPostRequestEncoder.bodyListDatas_ won't be cleaned.
So, the proposed fix is also cleaning everything from the _HttpPostRequestEncoder.bodyListDatas_ list.
Notice that form attributes are stored in both _HttpPostRequestEncoder.bodyListDatas_ list and in _DefaultHttpDataFactory.requestFileDeleteMap_ map, that's why the patch closes resources from _HttpPostRequestEncoder.bodyListDatas_ only if they are still accessible.

Example use case, let's say we do:

```
	@Test
	void _postUploadWithMultipart() throws Exception {
		Path file = Paths.get(getClass().getResource("/smallFile.txt").toURI());
		try (InputStream f = Files.newInputStream(file)) {
			doTestPostUpload((req, form) -> form.multipart(true)
							.file("test", f),
					"test ");
		}
	}
```

then when the HttpClientEncoder.run() method is executing, we have the following objects in _DefaultHttpDataFactory.requestFileDeleteMap_ and in _HttpPostRequestEncoder.bodyListDatas_ (the MemoryFileUpload is only available in _HttpPostRequestEncoder.bodyListDatas_, not in _DefaultHttpDataFactory.requestFileDeleteMap_:

![Screenshot 2022-10-25 at 16 54 03](https://user-images.githubusercontent.com/650758/197808200-5baab8ec-c421-4cb4-97d7-6c9a368779e4.png)

without the patch, the test will produce a leak error like this:

```
16:55:57.579 [Cleaner-0] ERROR io.netty5.buffer.LoggingLeakCallback - LEAK: Object "buffer (95 bytes)" was not property closed before it was garbage collected. A life-cycle back-trace (if any) is attached as suppressed exceptions. See https://netty.io/wiki/reference-counted-objects.html for more information.
io.netty5.buffer.LoggingLeakCallback$LeakReport: Object life-cycle trace:
	Suppressed: io.netty5.buffer.internal.LifecycleTracer$Traceback: ALLOCATE (current acquires = 0) T-73363us.
		at io.netty5.buffer.internal.ResourceSupport.<init>(ResourceSupport.java:42)
		at io.netty5.buffer.internal.AdaptableBuffer.<init>(AdaptableBuffer.java:28)
		at io.netty5.buffer.bytebuffer.NioBuffer.<init>(NioBuffer.java:65)
		at io.netty5.buffer.bytebuffer.ByteBufferMemoryManager.createBuffer(ByteBufferMemoryManager.java:85)
		at io.netty5.buffer.bytebuffer.ByteBufferMemoryManager.recoverMemory(ByteBufferMemoryManager.java:80)
		at io.netty5.buffer.pool.PooledBufferAllocator.allocate(PooledBufferAllocator.java:321)
		at io.netty5.buffer.DefaultBufferAllocators$UncloseableBufferAllocator.allocate(DefaultBufferAllocators.java:122)
		at io.netty.contrib.handler.codec.http.multipart.AbstractMemoryHttpData.setContent(AbstractMemoryHttpData.java:83)
		at reactor.netty5.http.client.HttpClientFormEncoder.file(HttpClientFormEncoder.java:203)
		at reactor.netty5.http.client.HttpClientForm.file(HttpClientForm.java:131)
		at reactor.netty5.http.client.HttpClientFormEncoder.file(HttpClientFormEncoder.java:147)
		at reactor.netty5.http.client.HttpClientWithTomcatTest.lambda$_postUploadWithMultipart$2(HttpClientWithTomcatTest.java:113)
		at reactor.netty5.http.client.HttpClientOperations$SendForm._subscribe(HttpClientOperations.java:841)
		at reactor.netty5.http.client.HttpClientOperations$SendForm.subscribe(HttpClientOperations.java:822)
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
		at reactor.netty5.http.client.HttpClientConnect$HttpIOHandlerObserver.onStateChange(HttpClientConnect.java:370)
		at reactor.netty5.ReactorNetty$CompositeConnectionObserver.onStateChange(ReactorNetty.java:653)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$DisposableAcquire.onStateChange(DefaultPooledConnectionProvider.java:188)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$DisposableAcquire.run(DefaultPooledConnectionProvider.java:250)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$DisposableAcquire.onNext(DefaultPooledConnectionProvider.java:172)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$DisposableAcquire.onNext(DefaultPooledConnectionProvider.java:100)
		at reactor.netty5.internal.shaded.reactor.pool.AbstractPool$Borrower.deliver(AbstractPool.java:467)
		at reactor.netty5.internal.shaded.reactor.pool.SimpleDequePool.lambda$drainLoop$9(SimpleDequePool.java:421)
		at reactor.core.publisher.FluxDoOnEach$DoOnEachSubscriber.onNext(FluxDoOnEach.java:154)
		at reactor.core.publisher.MonoCreate$DefaultMonoSink.success(MonoCreate.java:172)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$PooledConnectionAllocator$PooledConnectionInitializer.onNext(DefaultPooledConnectionProvider.java:551)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$PooledConnectionAllocator$PooledConnectionInitializer.onNext(DefaultPooledConnectionProvider.java:512)
		at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondComplete(MonoFlatMap.java:245)
		at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:305)
		at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
		at reactor.netty5.transport.TransportConnector$MonoChannelPromise.setSuccess(TransportConnector.java:483)
		at reactor.netty5.transport.TransportConnector.lambda$doConnect$8(TransportConnector.java:254)
		at io.netty5.util.concurrent.DefaultFutureListeners.notifyListeners(DefaultFutureListeners.java:77)
		at io.netty5.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:455)
		at io.netty5.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:440)
		at io.netty5.util.concurrent.DefaultPromise$NotifyListeners.run(DefaultPromise.java:423)
		at io.netty5.util.concurrent.SingleThreadEventExecutor.runTask(SingleThreadEventExecutor.java:338)
		at io.netty5.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:361)
		at io.netty5.channel.SingleThreadEventLoop.run(SingleThreadEventLoop.java:180)
		at io.netty5.util.concurrent.SingleThreadEventExecutor.lambda$doStartThread$4(SingleThreadEventExecutor.java:774)
		at io.netty5.util.internal.ThreadExecutorMap.lambda$apply$1(ThreadExecutorMap.java:68)
		at io.netty5.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		at java.base/java.lang.Thread.run(Thread.java:833)
	Suppressed: io.netty5.buffer.internal.LifecycleTracer$Traceback: SPLIT (split into two) T-40333us.
		at io.netty5.buffer.internal.CleanerDrop$GatedRunner.prepareRecover(CleanerDrop.java:130)
		at io.netty5.buffer.internal.CleanerDrop.attach(CleanerDrop.java:69)
		at io.netty5.buffer.internal.CleanerDrop.attach(CleanerDrop.java:31)
		at io.netty5.buffer.bytebuffer.NioBuffer.split(NioBuffer.java:501)
		at io.netty5.buffer.Buffer.readSplit(Buffer.java:872)
		at io.netty.contrib.handler.codec.http.multipart.AbstractMemoryHttpData.getChunk(AbstractMemoryHttpData.java:234)
		at io.netty.contrib.handler.codec.http.multipart.HttpPostRequestEncoder.encodeNextChunkMultipart(HttpPostRequestEncoder.java:893)
		at io.netty.contrib.handler.codec.http.multipart.HttpPostRequestEncoder.nextChunk(HttpPostRequestEncoder.java:1082)
		at io.netty.contrib.handler.codec.http.multipart.HttpPostRequestEncoder.readChunk(HttpPostRequestEncoder.java:1034)
		at reactor.netty5.http.client.HttpClientFormEncoder.readChunk(HttpClientFormEncoder.java:90)
		at reactor.netty5.http.client.HttpClientFormEncoder.readChunk(HttpClientFormEncoder.java:48)
		at io.netty5.handler.stream.ChunkedWriteHandler.doFlush(ChunkedWriteHandler.java:220)
		at io.netty5.handler.stream.ChunkedWriteHandler.flush(ChunkedWriteHandler.java:124)
		at io.netty5.channel.DefaultChannelHandlerContext.invokeFlush(DefaultChannelHandlerContext.java:892)
		at io.netty5.channel.DefaultChannelHandlerContext.invokeWriteAndFlush(DefaultChannelHandlerContext.java:907)
		at io.netty5.channel.DefaultChannelHandlerContext.write(DefaultChannelHandlerContext.java:923)
		at io.netty5.channel.DefaultChannelHandlerContext.writeAndFlush(DefaultChannelHandlerContext.java:902)
		at io.netty5.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:919)
		at io.netty5.channel.Channel.writeAndFlush(Channel.java:311)
		at reactor.netty5.http.client.HttpClientOperations$SendForm._subscribe(HttpClientOperations.java:879)
		at reactor.netty5.http.client.HttpClientOperations$SendForm.subscribe(HttpClientOperations.java:822)
		at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
		at reactor.netty5.http.client.HttpClientConnect$HttpIOHandlerObserver.onStateChange(HttpClientConnect.java:370)
		at reactor.netty5.ReactorNetty$CompositeConnectionObserver.onStateChange(ReactorNetty.java:653)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$DisposableAcquire.onStateChange(DefaultPooledConnectionProvider.java:188)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$DisposableAcquire.run(DefaultPooledConnectionProvider.java:250)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$DisposableAcquire.onNext(DefaultPooledConnectionProvider.java:172)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$DisposableAcquire.onNext(DefaultPooledConnectionProvider.java:100)
		at reactor.netty5.internal.shaded.reactor.pool.AbstractPool$Borrower.deliver(AbstractPool.java:467)
		at reactor.netty5.internal.shaded.reactor.pool.SimpleDequePool.lambda$drainLoop$9(SimpleDequePool.java:421)
		at reactor.core.publisher.FluxDoOnEach$DoOnEachSubscriber.onNext(FluxDoOnEach.java:154)
		at reactor.core.publisher.MonoCreate$DefaultMonoSink.success(MonoCreate.java:172)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$PooledConnectionAllocator$PooledConnectionInitializer.onNext(DefaultPooledConnectionProvider.java:551)
		at reactor.netty5.resources.DefaultPooledConnectionProvider$PooledConnectionAllocator$PooledConnectionInitializer.onNext(DefaultPooledConnectionProvider.java:512)
		at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondComplete(MonoFlatMap.java:245)
		at reactor.core.publisher.MonoFlatMap$FlatMapInner.onNext(MonoFlatMap.java:305)
		at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
		at reactor.netty5.transport.TransportConnector$MonoChannelPromise.setSuccess(TransportConnector.java:483)
		at reactor.netty5.transport.TransportConnector.lambda$doConnect$8(TransportConnector.java:254)
		at io.netty5.util.concurrent.DefaultFutureListeners.notifyListeners(DefaultFutureListeners.java:77)
		at io.netty5.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:455)
		at io.netty5.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:440)
		at io.netty5.util.concurrent.DefaultPromise$NotifyListeners.run(DefaultPromise.java:423)
		at io.netty5.util.concurrent.SingleThreadEventExecutor.runTask(SingleThreadEventExecutor.java:338)
		at io.netty5.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:361)
		at io.netty5.channel.SingleThreadEventLoop.run(SingleThreadEventLoop.java:180)
		at io.netty5.util.concurrent.SingleThreadEventExecutor.lambda$doStartThread$4(SingleThreadEventExecutor.java:774)
		at io.netty5.util.internal.ThreadExecutorMap.lambda$apply$1(ThreadExecutorMap.java:68)
		at io.netty5.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		at java.base/java.lang.Thread.run(Thread.java:833)
		Suppressed: io.netty5.buffer.internal.LifecycleTracer$Traceback: SPLIT (split from other object) T-40333us.
			... 50 common frames omitted
```

Reply:
I will do it soon, thanks.

Reply:
Applied the better patch from #2556 in commit [0f62d3c](https://github.com/reactor/reactor-netty/pull/2327/commits/0f62d3cb2d421b037c96c6850e19f60d3b676bbd).